File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq

def solution(operations):
    
    max_heap = []
    min_heap = []
    for i, op in enumerate(operations):
        o, n = op.split()
        n = int(n)
        
        if o == 'I':
            heapq.heappush(max_heap, -n)
            heapq.heappush(min_heap, n)
        else:
            # print(i, min_heap, max_heap)
            if not max_heap:
                continue
            if n == 1:
                max_value = -heapq.heappop(max_heap)
                min_heap.remove(max_value)
            else:
                min_value = heapq.heappop(min_heap)
                max_heap.remove(-min_value)
    if not max_heap:
        return [0, 0]
    return [-heapq.heappop(max_heap), heapq.heappop(min_heap)]
                

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.PriorityQueue;
import java.util.Collections;
import java.util.HashMap;

class Solution {
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> minPQ = new PriorityQueue<>();
        PriorityQueue<Integer> maxPQ = new PriorityQueue<>(Collections.reverseOrder());
        int count = 0;

        for (String operation : operations) {
            String[] ops = operation.split(" ");
            int num = Integer.parseInt(ops[1]);

            if (ops[0].equals("I")) {
                minPQ.add(num);
                maxPQ.add(num);
                count++;
            } else if (ops[0].equals("D")) {
                if (count == 0) continue;
                
                if (num == 1) {
                    minPQ.remove(maxPQ.poll());
                } else {
                    maxPQ.remove(minPQ.poll());
                }
                count--;
            }
        }

        if (count == 0) return new int[]{0, 0};

        int max = maxPQ.poll();
        int min = minPQ.poll();

        return new int[]{max, min};
    }
}
